NerdyPy/utils/download.py

```diff
@@ -1,87 +1,98 @@
-"""
-download method for Audio Content
-
-Needs testing!
-"""
-
-import os
-import shutil
-import urllib.request
-import uuid
-import ffmpeg
-from utils.errors import NerpyException
-
-import youtube_dl
-
-DL_DIR = "tmp"
-if not os.path.exists(DL_DIR):
-    os.makedirs(DL_DIR)
-
-YTDL_ARGS = {
-    "format": "bestaudio/best",
-    "outtmpl": os.path.join(DL_DIR, "%(id)s"),
-    "quiet": True,
-    "extractaudio": True,
-    "audioformat": "mp3",
-    "default_search": "auto",
-}
-
-
-def convert(file):
-    """Convert downloaded file to playable ByteStream"""
-    # TODO: Add better Exception handling
-    stream, _ = (
-        ffmpeg.input(file)
-        .filter("loudnorm")
-        .output("pipe:", format="mp3", ac=2, ar="48000")
-        .overwrite_output()
-        .run(capture_stdout=True)
-    )
-    return stream
-
-
-def download(url: str):
-    """download audio content (maybe transform?)"""
-    dlfile = ""
-    ytdl = youtube_dl.YoutubeDL(YTDL_ARGS)
-
-    req = urllib.request.Request(
-        url,
-        headers={
-            "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_9_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/35.0.1916.47 Safari/537.36"
-        },
-    )
-
-    split = os.path.splitext(url)
-
-    if split[1] is not None and split[1] != "":
-        dlfile = os.path.join(DL_DIR, f"{str(uuid.uuid4())}{split[1]}")
-        with urllib.request.urlopen(req) as response, open(dlfile, "wb") as out_file:
-            shutil.copyfileobj(response, out_file)
-    else:
-        video = ytdl.extract_info(url)
-        song = Song(**video)
-
-        for file in os.listdir(f"{DL_DIR}"):
-            if file.startswith(song.idn):
-                dlfile = os.path.join(DL_DIR, file)
-                break
-
-    if dlfile is None:
-        raise NerpyException(f"could not find a download in: {url}")
-    else:
-        return convert(dlfile)
-
-
-class Song:
-    """Song Model for YTDL"""
-
-    def __init__(self, **kwargs):
-        self.__dict__ = kwargs
-        self.title = kwargs.pop("title", None)
-        self.idn = kwargs.pop("id", None)
-        self.url = kwargs.pop("url", None)
-        self.webpage_url = kwargs.pop("webpage_url", "")
-        self.duration = kwargs.pop("duration", 60)
-        self.start_time = kwargs.pop("start_time", None)
-        self.end_time = kwargs.pop("end_time", None)
+"""
+download method for Audio Content
+
+Needs testing!
+"""
+
+import os
+import shutil
+import urllib.request
+import uuid
+import ffmpeg
+from utils.errors import NerpyException
+
+import youtube_dl
+
+DL_DIR = "tmp"
+if not os.path.exists(DL_DIR):
+    os.makedirs(DL_DIR)
+
+YTDL_ARGS = {
+    "format": "bestaudio/best",
+    "outtmpl": os.path.join(DL_DIR, "%(id)s"),
+    "quiet": True,
+    "extractaudio": True,
+    "audioformat": "mp3",
+    "default_search": "auto",
+}
+
+YTDL = youtube_dl.YoutubeDL(YTDL_ARGS)

```

Comment:
So wird  YTDL direkt beim start initialisiert, glaub ich? Was würdest du davon halten wenn wir hier ein lazy init machen?

Reply:
Geht das außerhalb von Klassen?

Reply:
`ytdl = None
def get_ytdl()
    if ytdl is None
        ytdl = youtube_dl.YoutubeDL(YTDL_ARGS)
    return ytdl
`

so oder so ähnlich

Reply:
Das sieht mir viel aufwändiger aus, als einfach nur diese eine Zeile zu haben? Bringt das einen Vorteil mit sich?

Reply:
jain. verhindert eigentlich nur, dass ytdl zum bot start initialisiert wird. Ist aber vmtl nicht so teuer, sodass sich das lohnen würde.